// react-scroll-to-bottom-updated/lib/BasicScrollToBottom.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports["default"] = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _AutoHideFollowButton = _interopRequireDefault(require("./ScrollToBottom/AutoHideFollowButton"));

var _Composer = _interopRequireDefault(require("./ScrollToBottom/Composer"));

var _Panel = _interopRequireDefault(require("./ScrollToBottom/Panel"));

var _useStyleToClassName = _interopRequireDefault(require("./hooks/internal/useStyleToClassName"));

var ROOT_STYLE = {
  position: 'relative'
};

var BasicScrollToBottomCore = function BasicScrollToBottomCore(_ref) {
  var children = _ref.children,
      className = _ref.className,
      followButtonClassName = _ref.followButtonClassName,
      scrollViewClassName = _ref.scrollViewClassName;
  var rootCSS = (0, _useStyleToClassName["default"])()(ROOT_STYLE);
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])(rootCSS, (className || '') + '')
  }, /*#__PURE__*/_react["default"].createElement(_Panel["default"], {
    className: (scrollViewClassName || '') + ''
  }, children), /*#__PURE__*/_react["default"].createElement(_AutoHideFollowButton["default"], {
    className: (followButtonClassName || '') + ''
  }));
};

BasicScrollToBottomCore.propTypes = {
  children: _propTypes["default"].any,
  className: _propTypes["default"].string,
  followButtonClassName: _propTypes["default"].string,
  scrollViewClassName: _propTypes["default"].string
};

var BasicScrollToBottom = function BasicScrollToBottom(_ref2) {
  var checkInterval = _ref2.checkInterval,
      children = _ref2.children,
      className = _ref2.className,
      debounce = _ref2.debounce,
      debug = _ref2.debug,
      followButtonClassName = _ref2.followButtonClassName,
      _ref2$initialScrollBe = _ref2.initialScrollBehavior,
      initialScrollBehavior = _ref2$initialScrollBe === void 0 ? 'smooth' : _ref2$initialScrollBe,
      mode = _ref2.mode,
      nonce = _ref2.nonce,
      scroller = _ref2.scroller,
      scrollViewClassName = _ref2.scrollViewClassName;
  return /*#__PURE__*/_react["default"].createElement(_Composer["default"], {
    checkInterval: checkInterval,
    debounce: debounce,
    debug: debug,
    initialScrollBehavior: initialScrollBehavior,
    mode: mode,
    nonce: nonce,
    scroller: scroller
  }, /*#__PURE__*/_react["default"].createElement(BasicScrollToBottomCore, {
    className: className,
    followButtonClassName: followButtonClassName,
    scrollViewClassName: scrollViewClassName
  }, children));
};

BasicScrollToBottom.propTypes = {
  checkInterval: _propTypes["default"].number,
  children: _propTypes["default"].any,
  className: _propTypes["default"].string,
  debounce: _propTypes["default"].number,
  debug: _propTypes["default"].bool,
  followButtonClassName: _propTypes["default"].string,
  initialScrollBehavior: _propTypes["default"].oneOf(['auto', 'smooth']),
  mode: _propTypes["default"].oneOf(['bottom', 'top']),
  nonce: _propTypes["default"].string,
  scroller: _propTypes["default"].func,
  scrollViewClassName: _propTypes["default"].string
};
var _default = BasicScrollToBottom;
exports["default"] = _default;
